### ISSUE TITLE
Sliced Score Matching: A Scalable Approach to Density and Score Estimation

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#205]
# TL; DR

- Project model score to lower dimensional space with random vectors for scalable optimization.

# Motivation

- Traditional score matching methods applies the integration by parts technique on Fisher divergence objective that mandates the computation of Hessian trace, which has linear time complexity w.r.t. the data dimension (*definitely not scalable*).
- Score estimation can empower many other frameworks that involve implicit density (such as implicit VAEs).

# Formulation

- Instead of computing the Fisher divergence between data score and model score, first project them to random directions using a set of random vectors.

![image](https://user-images.githubusercontent.com/46308407/212881227-9b471dba-a458-454b-a45a-46670e1050a3.png)
![image](https://user-images.githubusercontent.com/46308407/212881275-e0aea6e2-e5b1-47a6-9eb6-d73e30005857.png)

- Likewise, after integration by parts, the objective is altered from former to latter:

![image](https://user-images.githubusercontent.com/46308407/212881502-627f3bd1-f47e-4d44-b323-0b0c79fa8413.png)
![image](https://user-images.githubusercontent.com/46308407/212881545-78678ab9-db4f-480b-94fd-c0611e0f9024.png)

- In practice, of course, approximate the expectation with samples.

![image](https://user-images.githubusercontent.com/46308407/212881807-e7811003-6506-4891-bb31-0169ed894b15.png)

- If we sample ***v*** from MVN or Rademacher distribution, the second term can be integrated analytically, possibly reducing the variance of the estimation.

![image](https://user-images.githubusercontent.com/46308407/212882001-90a26eff-1646-4b9e-b456-fd30e390d3d7.png)

- Assuming we use *M* random directions for projection, we need ***M+1*** backprops instead of previous ***d+1*** (where *d* is the data dimension).
- The authors empirically find out that even ***M=1*** gives reasonable results.

![image](https://user-images.githubusercontent.com/46308407/212882400-87648eac-fe36-4546-b1eb-5b407e791341.png)

- For implementation, parameterize the score model ***h*** with a neural network and optimize the below objective for \theta.

![image](https://user-images.githubusercontent.com/46308407/212883576-1cb384d5-3dcd-479c-b0fc-691f45370f06.png)

- As we obtain the following equivalence, we can be assured that we are optimizing the minimize the projected difference between the score model and data score.
- It is interesting to note that with random vectors of sufficiently small norms, optimizing with SSM is equivalent to NCE.

![image](https://user-images.githubusercontent.com/46308407/212884438-297c2ae5-6a97-4834-9e5d-1b5edb8d8c57.png)
![image](https://user-images.githubusercontent.com/46308407/212884473-d9285ebc-ebff-4b98-934c-b52338ee9183.png)

# Evaluation

### Density Estimation

![image](https://user-images.githubusercontent.com/46308407/212884555-e3c50ac8-055b-4b50-90d9-063e28e20f3a.png)
![image](https://user-images.githubusercontent.com/46308407/212884588-e784d1e3-bb46-433c-94e5-928c424c11b2.png)
![image](https://user-images.githubusercontent.com/46308407/212884650-5b798104-9d4d-4aa7-b5af-68b865353717.png)

### Score Estimation (for implicit density models)

![image](https://user-images.githubusercontent.com/46308407/212884832-6d8043f2-bb42-4dfa-9574-2d356f672bfc.png)

- *It is a meaningful improvement, but still not quite there yet.*

# Inspiration

- Great introduction and backgrounds. I was relatively new to score-based models, and the authors gave me a very intuitive overview of the related literature.
- It is very interesting how optimizing divergences in randomly projected space yields similar results even with as little as 1 random vector.
